File: src/stmt.cpp
#include "stmt.hpp"

int Stmt::count = 0;

#include "garbage_collector.hpp"
#include "literal.hpp"

Stmt::Stmt() {
  count++;
  GarbageCollector<Stmt>::Push(this);
}

Stmt::Stmt(int ln) {
  line = ln;
  count++;
  GarbageCollector<Stmt>::Push(this);
}

Stmt::~Stmt() {
  count--;
  GarbageCollector<Stmt>::Pop(this);
}

void Stmt::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}
Block::Block(int ln, std::vector<Stmt*> stmtVector) {
    this->line = ln;
    this->stmtVector = stmtVector;
  }

Block::~Block() {
}

void Block::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

Break::Break(int ln) {
    this->line = ln;
  }

Break::~Break() {
}

void Break::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

Continue::Continue(int ln) {
    this->line = ln;
  }

Continue::~Continue() {
}

void Continue::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

Expression::Expression(int ln, Expr* expr) {
    this->line = ln;
    this->expr = expr;
  }

Expression::~Expression() {
}

void Expression::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

If::If(int ln, Expr* condition, Stmt* thenBranch, Stmt* elseBranch) {
    this->line = ln;
    this->condition = condition;
    this->thenBranch = thenBranch;
    this->elseBranch = elseBranch;
  }

If::~If() {
}

void If::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

Return::Return(int ln, Expr* result) {
    this->line = ln;
    this->result = result;
  }

Return::~Return() {
}

void Return::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

Use::Use(int ln, Token command) {
    this->line = ln;
    this->command = command;
  }

Use::~Use() {
}

void Use::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

Var::Var(int ln, Token name, Expr* initializer) {
    this->line = ln;
    this->name = name;
    this->initializer = initializer;
  }

Var::~Var() {
}

void Var::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

While::While(int ln, Expr* condition, Stmt* branch) {
    this->line = ln;
    this->condition = condition;
    this->branch = branch;
  }

While::~While() {
}

void While::Accept(StmtVisitor* visitor) {
  visitor->Visit(this);
}

File: src/Stmt.java
package com.krgamestudios.toy;

import java.util.List;

abstract class Stmt {
	interface Visitor<R> {
		R visit(Block stmt);
		R visit(Break stmt);
		R visit(Expression stmt);
		R visit(If stmt);
		R visit(Log stmt);
		R visit(Var stmt);
		R visit(While stmt);
	}

	abstract <R> R accept(Visitor<R> visitor);
	static class Block extends Stmt {
		final List<Stmt> statements;

		Block(List<Stmt> statements) {
			this.statements = statements;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Break extends Stmt {

		Break() {
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Expression extends Stmt {
		final Expr expression;

		Expression(Expr expression) {
			this.expression = expression;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class If extends Stmt {
		final Expr condition;
		final Stmt thenBranch;
		final Stmt elseBranch;

		If(Expr condition, Stmt thenBranch, Stmt elseBranch) {
			this.condition = condition;
			this.thenBranch = thenBranch;
			this.elseBranch = elseBranch;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Log extends Stmt {
		final Expr expression;

		Log(Expr expression) {
			this.expression = expression;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class Var extends Stmt {
		final Token name;
		final Expr initializer;

		Var(Token name, Expr initializer) {
			this.name = name;
			this.initializer = initializer;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
	static class While extends Stmt {
		final Expr condition;
		final Stmt body;

		While(Expr condition, Stmt body) {
			this.condition = condition;
			this.body = body;
		}

		<R> R accept(Visitor<R> visitor) {
			return visitor.visit(this);
		}
	}
}
